lib/tackle/consumer/executor.ex

```diff
@@ -21,47 +21,104 @@ defmodule Tackle.Consumer.Executor do
+  end
 
-    # start actual consuming
-    {:ok, _consumer_tag} = AMQP.Basic.consume(channel, topology.queue)
+  def handle_continue(:try_setup, %{channel_retry_ref: ref} = state) when is_reference(ref) do
+    {:noreply, state}
+  end
 
-    # Sleep is needed for the ConnectionErrorsTest
-    Process.sleep(2)
+  def handle_continue(
+        :try_setup,
+        %{
+          connection_id: connection_name,
+          rabbitmq_url: rabbitmq_url,
+          prefetch_count: prefetch_count,
+          topology: topology,
+          reconnect_interval: reconnect_interval,
+          handler: handler_module
+        } = state
+      ) do
+    with {:ok, connection} <- Tackle.Connection.open(connection_name, rabbitmq_url),
+         {:ok, channel} <- Tackle.Channel.create(connection, prefetch_count) do
+      # Get notifications when the connection goes down
+      Process.monitor(channel.pid)
+      Tackle.Consumer.Topology.setup!(channel, topology)
 
-    Tackle.Consumer.State.started(state, channel: channel)
+      {:noreply, %{state | channel: channel, channel_retry_ref: nil}, {:continue, :try_consume}}
+    else
+      {:error, _} = error ->
+        Logger.error("#{handler_module} failed to setup channel due to `#{inspect(error)}`")
+        ref = Process.send_after(self(), :setup_retry_timeout, reconnect_interval)
+        {:noreply, %{state | channel_retry_ref: ref}}
+    end
   end
 
-  defp setup_connection_channel(connection_id, url, prefetch_count) do
-    with {:ok, connection} <- Tackle.Connection.open(connection_id, url) do
-      # Get notifications when the connection goes down
-      Process.monitor(connection.pid)
-      channel = Tackle.Channel.create(connection, prefetch_count)
-      {:ok, channel}
+  # Try to start consumption
+  def handle_continue(:try_consume, %{consume_retry_ref: nil} = state) do
+    %{
+      channel: channel,
+      topology: topology,
+      reconnect_interval: reconnect_interval,
+      handler: handler_module
+    } = state
+
+    with {:ok, consumer_tag} <- AMQP.Basic.consume(channel, topology.queue) do
+      {:noreply, %{state | consumer_tag: consumer_tag}}
+    else
+      {:error, _} = error ->
+        Logger.error(
+          "#{handler_module} failed to start consumption from `#{topology.queue}` due to `#{inspect(error)}`"
+        )
+
+        ref = Process.send_after(self(), :consume_retry_timeout, reconnect_interval)
+        {:noreply, %{state | consume_retry_ref: ref}}
     end
   end
 
-  # Close channel on exit
-  def terminate(_reason, state) do
-    state.channel |> Tackle.Channel.close()
+  def handle_continue(:try_consume, state), do: {:noreply, state}
+
+  # Retry opening channel and setup topology
+  def handle_info(:setup_retry_timeout, state = %{channel_retry_ref: nil}), do: {:noreply, state}
+
+  def handle_info(:setup_retry_timeout, state = %{channel_retry_ref: ref})
+      when is_reference(ref) do
+    {:noreply, %{state | channel_retry_ref: nil}, {:continue, :try_setup}}
```

Comment:
A circuit breaker pattern or at least exponential back-off would be nice.
Otherwise I am a little worried what infinite retries might cause.

Reply:
@kaikuchn
We have an app with nearly 30 event consumers. I shut down RabbitMQ, and nothing happened. You get 30 "cannot connect" messages per second and nothing else. Just try it.

In my opinion, a circuit breaker is not needed here. This pattern is used to not overflow a process, e. g. when a process has a queue in front of it (letter box for elixir processes, a load balancer, or similar). A circuit breaker is used to stop putting more and more messages into this queue so that when the process gets responsive again, it wouldn't have a nearly endless queue to process and thus would basically become unresponsive again.
In our case, I cannot see how you can get more than the number of consumers asking at the same time for a connection, since all operations(`Tackle.Connection.open`, `Tackle.Channel.open` and `AMQP.Basic.consume`) are synchronous... I could add a couple more tests for those cases. Turning off RabbitMQ is easy, you just get an immediate `{:error, :connection_refused}` or similar. Simulating the starting process of RabbitMQ, so basically running into a timeout might be more tricky.

Reply:
You're right, a circuit breaker pattern is unnecessary.

But I think an exponential back-off might be beneficial in some circumstances. I'm not very worried about completely unavailable RabbitMQ servers, but about other failures. E.g. partially/unreliably available server, misconfigured clients (e.g., connection succeeds, but channel creation fails due to wrong redeclaration, non existing queues, or similar). These failures with excessive re-attempts may lead to high churn rates, which may further exhaust resources on the server and lead to other failures.

One possible way to induce such a failure for channel creation would be binding to non existing queues.
Misconfigured TLS might be a good way to test reconnection attempts.

We've taken steps to mitigate such problems by configuring our RabbitMQ Server to have connection limits per user, and started using different users per app. So in our case this isn't a big problem, at worst the client will run into the connection limit but it won't take down the whole server for everyone else.

--> I think a back-off mechanism would be prudent, but I won't insist on it until I've seen it cause problems :)